### PR TITLE
more thread safety and events executor test

### PR DESCRIPTION
**Pull request by alsora**
Signed-off-by: Soragna, Alberto <alberto.soragna@gmail.com>

### CODE REVIEW

File: rclcpp/src/rclcpp/executors/timers_manager.cpp [hunk 1]
```diff
@@ -80,92 +96,134 @@ void TimersManager::stop()
 
 std::chrono::nanoseconds TimersManager::get_head_timeout()
 {
+  // Do not allow to interfere with the thread running
+  if (running_) {
```
Comment (mauropasse) [inline rclcpp/src/rclcpp/executors/timers_manager.cpp:100]: also here

File: rclcpp/src/rclcpp/executors/timers_manager.cpp [hunk 2]
```diff
@@ -80,92 +96,134 @@ void TimersManager::stop()
...
+  }
+
   std::unique_lock<std::mutex> lock(timers_mutex_);
   return this->get_head_timeout_unsafe();
 }
 
 void TimersManager::execute_ready_timers()
 {
+  // Do not allow to interfere with the thread running
+  if (running_) {
```
Comment (mauropasse) [inline rclcpp/src/rclcpp/executors/timers_manager.cpp:112]: and here

File: rclcpp/src/rclcpp/executors/timers_manager.cpp [hunk 3]
```diff
@@ -61,15 +67,25 @@ void TimersManager::start()
...
+  std::unique_lock<std::mutex> timers_lock(timers_mutex_);
   running_ = true;
   timers_thread_ = std::thread(&TimersManager::run_timers, this);
   pthread_setname_np(timers_thread_.native_handle(), "TimersManager");
 }
 
 void TimersManager::stop()
 {
-  // Notify stop to timers thread
+  // Nothing to do if the timers thread is not running
```
[the remark this answers is not in the document]
Comment (alsora) [reply rclcpp/src/rclcpp/executors/timers_manager.cpp:78]: Do we really need a mutex there?

File: rclcpp/src/rclcpp/executors/timers_manager.cpp [hunk 4]
```diff
@@ -80,92 +96,134 @@ void TimersManager::stop()
...
   this->execute_ready_timers_unsafe();
 }
 
-void TimersManager::execute_ready_timers_unsafe()
+bool TimersManager::execute_head_timer()
 {
-  if (heap_.empty()) {
-    return;
+  // Do not allow to interfere with the thread running
+  if (running_) {
```
Comment (mauropasse) [inline rclcpp/src/rclcpp/executors/timers_manager.cpp:124]: and here

**Review by irobot-ros — APPROVED** [review]

**Pull request merged**